packages/platforms/delta/x86-64/x86-64-delta-agc7648sv1/onlp/builds/src/module/src/sfpi.c

```diff
@@ -0,0 +1,467 @@
+/************************************************************
+ * <bsn.cl fy=2014 v=onl>
+ *
+ *           Copyright 2014 Big Switch Networks, Inc.
+ *           Copyright (C) 2017  Delta Networks, Inc. 
+ *
+ * Licensed under the Eclipse Public License, Version 1.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ *        http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the
+ * License.
+ *
+ * </bsn.cl>
+ ************************************************************
+ *
+ *
+ *
+ ***********************************************************/
+#include <onlp/platformi/sfpi.h>
+#include <fcntl.h> /* For O_RDWR && open */
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <math.h>
+#include <onlplib/file.h>
+#include <onlplib/i2c.h>
+#include "platform_lib.h"
+
+/******************* Utility Function *****************************************/
+int sfp_map_bus[] = {
+    51, 52, 53, 54, 55, 56, 57, 58, 59, 60, /* SFP */
+    61, 62, 63, 64, 65, 66, 67, 68, 69, 70, /* SFP */
+    71, 72, 73, 74, 75, 76, 77, 78, 79, 80, /* SFP */
+    81, 82, 83, 84, 85, 86, 87, 88, 89, 90, /* SFP */
+    91, 92, 93, 94, 95, 96, 97, 98, /* SFP */
+    41, 42, 43, 44, 45, 46, /* QSFP */
+};
+
+/************************************************************
+ *
+ * SFPI Entry Points
+ *
+ ***********************************************************/
+int onlp_sfpi_init(void)
+{
+    /* Called at initialization time */
+    lockinit();
+    return ONLP_STATUS_OK;
+}
+
+int onlp_sfpi_map_bus_index(int port)
+{
+    if(port < 0 || port > 54)
+        return ONLP_STATUS_E_INTERNAL;
+    return sfp_map_bus[port-1];
+}
+
+int onlp_sfpi_bitmap_get(onlp_sfp_bitmap_t* bmap)
+{
+    /* Ports {1, 54} */
+    int p;
+    AIM_BITMAP_CLR_ALL(bmap);
+
+    for(p = 1; p <= NUM_OF_PORT; p++) {
+        AIM_BITMAP_SET(bmap, p);
+    }
+    return ONLP_STATUS_OK;
+}
+  
+int onlp_sfpi_is_present(int port)
+{
+    char port_data[2];
```

Comment:
I optimized the third parameter of write() that is depend on character length to write data out of a buffer.

Reply:
That sounds great. Still the first part is not solved, e.g. in case port 42 is addressed `port_data` will contain just contain something like `{'4', '2'}` but you would like to have something that looks like `{'4', '2', '\0'}` (i.e. having the string termination). Otherwise all functions that work on that string (strlen, atoi, *strto*, ...)  depend on what is next in memory which will lead to an unwanted behaviour in case the next char in the memory is a valid string representation of a number.

Reply:
Add initial value for arrays to prevent unwanted behaviour.